### PR TITLE
Prepare for latest Docker Engine: Add npipe

### PR DESCRIPTION
**Pull request by StefanScherer**
After testing https://github.com/docker/docker/pull/21040 I found out that I have to add another option `-H npipe:////./pipe/docker_engine` to the Docker Engine to listen on the named pipe as well. See https://github.com/docker/docker/pull/19911 for the introduction of the npipe feature.

This PR is incompatible with the TP4 Docker Engine and should be merged after a new Windows Docker Engine is released.

I just want to write it down somewhere if someone else is trying the latest Docker Engine.

### CODE REVIEW

**Comment by enderb-ms** [comment]
@jhowardmsft

File: virtualization/windowscontainers/deployment/docker_windows.md [hunk 1]
```diff
@@ -42,11 +42,11 @@ mkdir %ProgramData%\docker
 :run
 if exist %certs%\server-cert.pem (goto :secure)
 
-docker daemon -D -b "Virtual Switch"
+docker daemon -D -b "Virtual Switch" -H npipe:////./pipe/docker_engine
```
Comment (jhowardmsft) [inline virtualization/windowscontainers/deployment/docker_windows.md]: This is not needed as the default is this.

File: virtualization/windowscontainers/management/container_networking.md [hunk 2]
```diff
@@ -124,7 +124,7 @@ Stop-Service docker
 The configuration file can be found at `c:\programdata\docker\runDockerDaemon.cmd`. Edit the following line, replacing `Virtual Switch` with the name of the virtual switch to be used by the Docker service.
 
 ```powershell
-docker daemon -D -b “New Switch Name"
+docker daemon -D -b “New Switch Name" -H npipe:////./pipe/docker_engine
```
Comment (jhowardmsft) [inline virtualization/windowscontainers/management/container_networking.md]: Shouldn't need a -H as this is the default.

File: virtualization/windowscontainers/deployment/docker_windows.md [hunk 3]
```diff
@@ -121,7 +121,7 @@ Download docker.exe from `https://aka.ms/tp4/docker` and copy it to the `windows
 Run the below command to start the docker daemon. This will need to be run each time the container host is started. This command starts the Docker daemon, specifies a virtual switch for container connectivity, and set’s the daemon to listen on port 2375 for incoming Docker requests. In this configuration Docker can be managed from a remote computer.
 
 ```powershell
-PS C:\> start-process cmd "/k docker daemon -D -b <Switch Name> -H 0.0.0.0:2375”
+PS C:\> start-process cmd "/k docker daemon -D -b <Switch Name> -H npipe:////./pipe/docker_engine -H 0.0.0.0:2375”
```
Comment (StefanScherer) [inline virtualization/windowscontainers/deployment/docker_windows.md]: Oh, just found "bad" typographic double quotes at the end of this line and some other places as well.

File: virtualization/windowscontainers/deployment/docker_windows.md [hunk 4]
```diff
@@ -169,7 +169,7 @@ Now, on your machine, open a PowerShell or CMD session, and run the Docker comma
 .\docker.exe -H tcp://<ip address of Nano Server>:2375
 ```
 
-For example, if you would like to see the available images: 
+For example, if you would like to see the available images:
 
 ```powershell
 .\docker.exe -H tcp://<ip address of Nano Server>:2375 images
```
Comment (StefanScherer) [inline virtualization/windowscontainers/deployment/docker_windows.md]: I think here a remote setup is wanted as the client runs on your machine and the Nano Server is another machine with different IP address.

**Comment by StefanScherer** [comment]
@jhowardmsft Thanks for the review. I've removed the uneccesary -H options. I've rebased the PR.

(My editor (Atom) found a lot of trailing white spaces and removed them.)

File: virtualization/windowscontainers/deployment/docker_windows.md (hunk 4 above)
Comment (jhowardmsft) [inline virtualization/windowscontainers/deployment/docker_windows.md]: That should still be fine - use npipe:////servername/pipe/docker_engine. There should not be any explicit 2375 listening on TCP regardless of nano/servercore, local or remote.

**Comment by scooley** [comment]
@StefanScherer the whitespace thing is such a mixed bag.  Sometimes it's unnecessary, sometimes it's two spaces to make a line break rather than paragraph break...

Atom is great about preserving the double spaces and removing the single spaces.  In short.  DOn't worry about it

File: virtualization/windowscontainers/management/container_networking.md [hunk 5]
```diff
@@ -124,7 +124,7 @@ Stop-Service docker
 The configuration file can be found at `c:\programdata\docker\runDockerDaemon.cmd`. Edit the following line, replacing `Virtual Switch` with the name of the virtual switch to be used by the Docker service.
 
 ```powershell
-docker daemon -D -b “New Switch Name"
+docker daemon -D -b "New Switch Name"
```
Comment (jhowardmsft) [inline virtualization/windowscontainers/management/container_networking.md]: Good catch

File: windows-server-container-tools/Install-ContainerHost/Install-ContainerHost.ps1 [hunk 6]
```diff
@@ -1168,12 +1168,12 @@ mkdir %ProgramData%\docker
...
 if exist %certs%\server-cert.pem (if exist %ProgramData%\docker\tag.txt (goto :secure))
- 
+
 docker daemon -D -b "$global:SwitchName"
 goto :eof
- 
+
 :secure
-docker daemon -D -b "$global:SwitchName" -H 0.0.0.0:2376 --tlsverify --tlscacert=%certs%\ca.pem --tlscert=%certs%\server-cert.pem --tlskey=%certs%\server-key.pem
+docker daemon -D -b "$global:SwitchName" -H npipe:////./pipe/docker_engine -H 0.0.0.0:2376 --tlsverify --tlscacert=%certs%\ca.pem --tlscert=%certs%\server-cert.pem --tlskey=%certs%\server-key.pem
```
Comment (jhowardmsft) [inline windows-server-container-tools/Install-ContainerHost/Install-ContainerHost.ps1]: Actually, thought about this some more. I think (not verified) that TLS will work over npipe, in which case the TCP listener shouldn't be required. Need someone like @taylorb-microsoft to verify if that is the case.

File: virtualization/windowscontainers/management/container_networking.md (hunk 5 above)
Comment (StefanScherer) [inline virtualization/windowscontainers/management/container_networking.md]: Just CMD+SHIFT+F (or CTRL+SHIFT+F) with the `“` to search through the whole repo. Should be done for `”` as well.

File: windows-server-container-tools/Install-ContainerHost/Install-ContainerHost.ps1 (hunk 6 above)
Comment (StefanScherer) [inline windows-server-container-tools/Install-ContainerHost/Install-ContainerHost.ps1]: I've tested this with @taylorb-microsoft's TLS script today morning and it worked fine. TLS over 2376 from outside and probably no TLS with npipe to work within the TP4 machine locally. Tested this with a Mac Docker Client against the TP4 VM.

Works just like the Linux version:

```
/usr/local/bin/docker daemon -D -g /var/lib/docker -H unix:// -H tcp://0.0.0.0:2376 --label provider=vmwarefusion --tlsverify --tlscacert=/var/lib/boot2docker/ca.pem --tlscert=/var/lib/boot2docker/server.pem --tlskey=/var/lib/boot2docker/server-key.pem -s aufs
```

Comment (StefanScherer) [inline windows-server-container-tools/Install-ContainerHost/Install-ContainerHost.ps1]: I will verify this again, tried this today morning and had a problem. That's why I search through the source and found the default pipe name.

Comment (jhowardmsft) [inline windows-server-container-tools/Install-ContainerHost/Install-ContainerHost.ps1]: It does.

```
E:\go\src\github.com\docker\docker\docker>docker daemon -H npipe://
time="2016-03-10T13:46:14.892344900-08:00" level=info msg="Graph migration to co
ntent-addressability took 0.00 seconds"
time="2016-03-10T13:46:14.892843200-08:00" level=info msg="Loading containers: s
tart."

time="2016-03-10T13:46:14.892843200-08:00" level=info msg="Loading containers: d
one."
time="2016-03-10T13:46:14.893343400-08:00" level=info msg="Daemon has completed
initialization"
time="2016-03-10T13:46:14.893343400-08:00" level=info msg="Docker daemon" commit
=c545f17-dirty graphdriver=windowsfilter version=1.11.0-dev
time="2016-03-10T13:46:14.898841600-08:00" level=info msg="API listen on //./pip
e/docker_engine"
```

Comment (StefanScherer) [inline windows-server-container-tools/Install-ContainerHost/Install-ContainerHost.ps1]: Confirmed. I'll simplify the PR.

**Comment by taylorb-microsoft** [comment]
Ignoring the white space the raw change is the same change that I was playing with...

I think @mebersol was also planning to make this change and you beat him to it :+1:

**Comment by mebersol** [comment]
Is this change backwards-compatible with TP4 Docker?

**Comment by StefanScherer** [comment]
@mebersol No, this is **not compatible** and should be merged after a new release is there. Probably the download link should be updated to make this clear (changing the line with `$DockerPath = "https://aka.ms/tp4/docker",`).

Using TLS with the (Nov 23) TP4 docker.exe also is a little tricky. As the old default was port 2375 w/o TLS you can't just add `-H tcp://127.0.0.1:2375 -H tcp://0.0.0.0:2376` because then local port 2375 also is TLS. Docker Engine only turns on and off TLS globally for all tcp ports.

So TLS makes sense with the `npipe://` feature to work both locally w/o TLS and remote with TLS.

**Comment by mebersol** [comment]
OK, thanks for the confirm - that's what I thought.

**Comment by scooley** [comment]
@mebersol @jhowardmsft -- reading the thread I'm not sure if I should merge into live or merge into TP5 staging.

**Comment by jhowardmsft** [comment]
TP5 staging

**Comment by neilpeterson** [comment]
@jhowardmsft -  This is what is working for me, can you verify that this is the expected configuration  and behavior?

Start Docker (non-tls):
```
docker daemon -D -H npipe:// -H 0.0.0.0:2375
```

Access Docker from remote system:
```
docker -H tcp://10.121.106.78:2375
```

**Comment by jhowardmsft** [comment]
All you're saying there is that the daemon is listening (securely) on npipe locally, and (insecurely) on all network interfaces over TCP.  I don't think we want to recommend insecure TCP configuration ever.

**Comment by neilpeterson** [comment]
Thanks - fair enough. So then the recommendations would be:

```
docker daemon -D
```
and / or
```
docker daemon -D -H npipe:// -H 0.0.0.0:2376 --tlsverify --tlscacert=%certs%\ca.pem --tlscert=%certs%\server-cert.pem --tlskey=%certs%\server-key.pem
```

Just want to validate that the `-H npipe://` is necessary. From my insecure tests, it is.

Thanks

**Comment by StefanScherer** [comment]
Yes. Having no -H option it defaults to npipe:// (as it would default to unix:// on Linux), and giving one -H for insecure/secure TCP connections you also need the second -H for npipe://.

One other detail. I compare the command lines with the Linux distros at the moment.
A boot2docker / docker-machine VM uses this command line:

```
/usr/local/bin/docker daemon -D -g /var/lib/docker -H unix:// -H tcp://0.0.0.0:2376 ...
``` 

So there is a `tcp://0.0.0.0:2376` instead of just `0.0.0.0:2376`.
Should we add the `tcp://` to be more consistent with the Linux docs?

**Comment by neilpeterson** [comment]
Thanks Stefan, agreed on `tcp://`. I've just tested, it works fine, I will adopt this for TP5 ++ docs.

neilp

**Comment by StefanScherer** [comment]
OK, I've fixed it in this PR as well.

**Comment by friism** [comment]
@jhowardmsft I just tried tooling around with mounting nipe into a windows container using `-v`. If one can negotiate the path-wrangling, do you think that's possible?

Mounting the Docker socket is common (for reasons good and bad) on Linux

**Comment by friism** [comment]
@taylorb-microsoft are you thinking about ways to make socket mounting happen?

**Comment by jhowardmsft** [comment]
Not technically possible. Windows is not Linux!

**Comment by friism** [comment]
Could we at least entertain workarounds...? We shouldn't bake it into the engine, but I wonder if there might be a way to proxy the HTTP API into a running container some other way.

**Comment by friism** [comment]
Could we at least entertain workarounds...? We shouldn't bake it into the engine, but I wonder if there might be a way to proxy the HTTP API into a running container some other way.

**Comment by jhowardmsft** [comment]
IF (capitals) there is, it's way down the list. If it requires platform support (it likely will), I am extremely doubtful there are resources to entertain working on this this side of RTM.

**Comment by taylorb-microsoft** [comment]
I want to make sure I understand the scenario for socket mounting...  Is the scenario for getting basic network access in the container w/out using NAT?

**Comment by friism** [comment]
@taylorb-microsoft no, it's mounting the docker engine API socket. That way you can run a docker container that can introspect itself and the other containers running on the host, and schedule new containers. Software like UCP uses this for installing itself.

**Comment by friism** [comment]
( @taylorb-microsoft again, this is non-urgent - I'm just interested in exploring what we could do for this )

**Comment by neilpeterson** [comment]
I believe these changes have been merged into today's release. I am going to close out this PR. If I've overlooked anything, just let me know.

**Comment by StefanScherer** [comment]
No problem. Yes the new runDockerDaemon.cmd looks good.

**Pull request closed without merging**